processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndexCursorHolder.java

```diff
@@ -73,10 +79,17 @@ public Cursor asCursor()
     return new IncrementalIndexCursor(
         rowSelector,
         spec,
-        Cursors.getTimeOrdering(ordering)
+        timeColumnName,
+        getTimeOrder(ordering, timeColumnName)
     );
   }
 
+  @Override
+  public boolean isPreAggregated()
+  {
+    return isPreaggregated;
+  }
+
```

Comment:
The flow is this, the query engines make a `CursorBuildSpec` that is given to the `CursorFactory` to produce a `CursorHolder`. If the `CursorBuildSpec` is aggregating, the `CursorFactory` can see if it has all of the required grouping and filtering columns, equivalent virtual columns (or all required inputs) and equivalent aggregators. If all of these things are true, then the `CursorFactory` returns a `CursorHolder` with `isPreAggregated` set to true, and the selector factories expose the aggregator columns as the output name of the agg factories on the `CursorBuildSpec` (it also does this for virtual columns if they are pre-existing and don't need recomputed). Because these agg columns were built by the same agg factory otherwise, the engine can safely check `isPreAggregated` and use the combining factory to process the segment instead.

I think this pattern should safe and has worked for everything I have tried, because its also how merging of per segment results works.

I couldn't find any cases where the aggregator renames its output when calling `getCombiningFactory` and i looked at all of the implementations. I'm not sure it would be allowed, because it would break any result format that was map based, so it seems to me that the output name must be stable for both the initial pass and also merging results from the per segment results with the combining aggregator. As far as I can tell, the factory stores its output at output name, so the combining factory will always make selectors that reads inputs from that output name to do segment merging.